Simplification of combinatorial expressions

mohit3011 opened the pull request:
Fixes #6341

Commit:
Add gammasimp in __init__.py

Commit:
Add gammasimp function in combsimp.py

Commit:
Add gammasimp test in tests_combsimp.py

Commit:
Update combsimp.py

Commit:
Update combsimp.py

Commit:
Merge remote-tracking branch 'upstream/master' into Fixes#6341

Commit:
Update combsimp.py

Commit:
Update combsimp.py

Commit:
Update combsimp.py

Commit:
Update combsimp.py

mohit3011 commented:
@jksuom  , please have a look in combsimp.py

Commit:
Merge remote-tracking branch 'upstream/master' into Fixes#6341

Commit:
Update combsimp.py

jksuom commented on sympy/simplify/combsimp.py:
```diff
@@ -66,9 +64,14 @@ def combsimp(expr):
+    as_gamma = expr.has(gamma)
+
+    if expr.has(gamma):
+        if not expr.has(factorial):
```
Maybe all expressions involving gamma should be dispatched to gammasimp regardless of whether they also have factorial (or binomial) or not, at least before the simplification of purely combinatorial expressions has been worked out. Then it would not be necessary to consider gamma in this function.

Commit:
Update combsimp.py

jksuom commented:
This is what Travis reports:

AssertionError: File contains trailing whitespace: /home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/sympy/simplify/combsimp.py, line 873.

Commit:
Merge remote-tracking branch 'upstream/master' into Fixes#6341

Commit:
Update combsimp.py

Commit:
Merge remote-tracking branch 'upstream/master' into Fixes#6341

Commit:
Reduce combsimp function

Commit:
Reduce combsimp function

Commit:
Reduce combsimp function

Commit:
Reduce combsimp function

mohit3011 commented:
@jksuom  , I tried to remove parts in combsimp function using gamma but the problem is that the code is interdependent and even removing certain portion gives failure in tests . I have made the provision of redirecting any function with gamma into gammasimp.

mohit3011 commented:
Is it okay if we leave combsimp function as it is for now and merge the other changes

Commit:
Modify combsimp.py

jksuom commented:
There should be tests showing that the issue has been fixed. That will probably involve more work. For example, it seems that gamma functions with integer arguments should be first converted to factorials and then handled by combsimp.

mohit3011 commented:
@jksuom , could you guide me what to do next

jksuom commented:
combsimp should be dedicated to functions of combinatorial type, i.e,. binomials and factorials, also rising and falling. The arguments of factorials should be integers. gamma functions with integer arguments can be converted to factorials and handled by combsimp. binomials and rising/falling factorials that have two arguments may be allowed non-integer first argument but the second argument should be an integer. Other types of components should be handled by gammasimp. In particular, those parts of the current combsimp that handle gammas with non-integer arguments should be omitted from combsimp (and left to gammasimp alone).

The dedicated combsimp should be able to fix #6341.

Commit:
Merge remote-tracking branch 'upstream/master' into Fixes#6341

Commit:
Update for non-integer argument in gammasimp

mohit3011 commented:
@jksuom  , I looked into the arguments being integer and I added these lines.

1. `if as_gamma and (not expr.args[0].is_Integer):` , this is for only sending those expressions in gammasimp which don't have an integer argument

2. `if as_factorial and expr.args[0].is_Integer` , this is for factorial should have integer argument

But how to test this , I mean is there any way to convert symbols in test_combsimp to integes and test them because without type casting they are giving error when I ran the local tests.

Commit:
Update for integer arguments

Commit:
Merge remote-tracking branch 'upstream/master' into Fixes#6341

jksuom commented on sympy/simplify/combsimp.py:
```diff
@@ -66,9 +64,12 @@ def combsimp(expr):
     as_binomial = expr.has(binomial)
+    as_gamma = expr.has(gamma)
+
+    if as_gamma and (not expr.args[0].is_Integer):
```
`expr.args[0]` is not the argument of the `expr.func` which is not `gamma`, in general. For example,
```
>>> expr = gamma(n + S(1)/3)*gamma(n + S(2)/3)
>>> expr.func
<class 'sympy.core.mul.Mul'>
>>> expr.args[0]
gamma(n + 1/3)
>>> expr.args[1]
gamma(n + 2/3)
```
To find the arguments it is necessary to examine each application of `gamma` more closely.

But that is not necessary to do now. It can be done as a final touch. At this point the arguments need not be inspected.

Commit:
Merge remote-tracking branch 'upstream/master' into Fixes#6341

Commit:
Remove functions involved in pure gamma work from combsimp

Commit:
Remove the tests for gamma functions of combsimp

mohit3011 commented:
@jksuom, I have removed two functions from combsimp which were involved purely in gamma calculations.  Furthermore, I have removed the test for gamma function in combsimp because same tests are there for gammasimp.

Commit:
Remove functions from combsimp.py

jksuom commented on sympy/simplify/combsimp.py:
```diff
@@ -217,6 +219,232 @@ def explicate(p):
             elif isg is False:
                 denom_others.extend(l)
+        # =========== level >= 2 work: factor absorbtion =========
+        if level >= 2:
```
It looks like this is a part of `combsimp`. Only `gammasimp` should be dealing with `gamma`. There could be something similar in `combsimp` but handling only factorials (possibly also rising and falling), not gammas.

mohit3011 replied:
@jksuom, removal of "level>=2 work : factor absorption" from combsimp is giving error. Also any similar function would contain gamma due to the fact that there maybe some nested gamma function which would be evaluated.
If there is any other way please suggest me. Thanks.

jksuom replied:
There should not be any gammas in the expression at this point. Which kind of error is arising?

mohit3011 replied:
When I removed If condition after line 223 i.e the whole if condition with its condition , it gave
![screen shot 2017-02-19 at 11 21 10 pm](https://cloud.githubusercontent.com/assets/19210197/23104783/57b25058-f6fa-11e6-9078-198ce6b86518.png)

jksuom replied:
Neither of these tests involve gammas. There should be new code dealing with pure combinatorial functions to implement these simplifications.

mohit3011 replied:
@jksuom I looked into "find_fuzzy" function and the definitions of all functions involved in it. But, I am not sure how to proceed to make it deal with pure combinatorial functions. I am not quite familiar with the functions in other modules, so please help me. Thanks

jksuom replied:
It may be too early to consider the workings of `find_fuzzy` at this point. There are a lot of other algorithms in `rule_gamma` that do not depend much on other modules. Perhaps you could try to understand what they do and how they should be modified when `gamma` is replaced by `factorial`. Some parts (for example, the one based on the formula `gamma(x)*gamma(1-x) = pi/sin(pi*x)`) do not apply to `factorial` and can be removed.

mohit3011 replied:
@jksuom , I have removed two functions from rule_gamma and changed one function for factorial purpose only . I have two doubts:
1) In the new factorial_factor function what does level 0,1 and 3 denote.
2) what does function explicate exactly do ?
Please explain , thanks.

Commit:
Merge remote-tracking branch 'upstream/master' into Fixes#6341

Commit:
Remove two functions related to gamma handling from rule_gamma

Commit:
Change gamma_factor to factorial_factor

Commit:
Merge remote-tracking branch 'upstream/master' into Fixes#6341

Commit:
Merge remote-tracking branch 'upstream/master' into Fixes#6341

Commit:
Merge remote-tracking branch 'upstream/master' into Fixes#6341

Commit:
Remove further gamma check in combsimp

Commit:
Minor changes

Commit:
Merge remote-tracking branch 'upstream/master' into Fixes#6341

Commit:
Merge remote-tracking branch 'upstream/master' into Fixes#6341

Commit:
Update for rewrite function

Commit:
Merge remote-tracking branch 'upstream/master' into Fixes#6341

mohit3011 commented:
@jksuom  , please review for new rewrite function

jksuom commented on sympy/functions/combinatorial/factorials.py:
```diff
@@ -852,7 +852,7 @@ def _eval_expand_func(self, **hints):
 
     def _eval_rewrite_as_factorial(self, n, k):
-        return factorial(n)/(factorial(k)*factorial(n - k))
+        return "binomial(n,k)"
```
This rewrite should not be changed. Only combsimp should be changed to rewrite this combination of factorials as a binomial.

mohit3011 replied:
Okay , thanks for the review

Commit:
Modifications

Commit:
Merge remote-tracking branch 'upstream/master' into Fixes#6341

Commit:
Merge remote-tracking branch 'upstream/master' into Fixes#6341

Commit:
Merge remote-tracking branch 'upstream/master' into Fixes#6341

Commit:
Merge remote-tracking branch 'upstream/master' into Fixes#6341

Commit:
Merge remote-tracking branch 'upstream/master' into Fixes#6341

Commit:
Merge remote-tracking branch 'upstream/master' into Fixes#6341

Commit:
Merge remote-tracking branch 'upstream/master' into Fixes#6341

ylemkimon commented:
Closing this, since #13210 has been merged.
